tests/test_convert_positional_args.py

```diff
@@ -0,0 +1,107 @@
+from typing import List
+
+import pytest
+
+from optuna._convert_positional_args import convert_positional_args
+
+
+def _sample_func(*, a: int, b: int, c: int) -> int:
+    return a + b + c
+
+
+def test_convert_positional_args_decorator() -> None:
+    previous_positional_arg_names: List[str] = []
+    decorator_converter = convert_positional_args(
+        previous_positional_arg_names=previous_positional_arg_names
+    )
+
+    decorated_func = decorator_converter(_sample_func)
+    assert decorated_func.__name__ == _sample_func.__name__
+
+
+def testconvert_positional_args_user_warning() -> None:
+    previous_positional_arg_names: List[str] = ["a", "b"]
+    decorator_converter = convert_positional_args(
+        previous_positional_arg_names=previous_positional_arg_names
+    )
+    assert callable(decorator_converter)
+
+    decorated_func = decorator_converter(_sample_func)
+    with pytest.warns(FutureWarning) as record:
+        decorated_func(1, 2, c=3)  # type: ignore
+        decorated_func(1, b=2, c=3)  # type: ignore
+        decorated_func(a=1, b=2, c=3)  # no warn
+
+    assert len(record) == 2
+    for warn in record.list:
+        assert isinstance(warn.message, FutureWarning)
+        assert _sample_func.__name__ in str(warn.message)
+
+
+def testconvert_positional_args_mypy_type_inference() -> None:
+    previous_positional_arg_names: List[str] = []
+    decorator_converter = convert_positional_args(
+        previous_positional_arg_names=previous_positional_arg_names
+    )
+    assert callable(decorator_converter)
+
+    class _Sample(object):
+        def __init__(self) -> None:
+            pass
+
+        def method(self) -> bool:
+            return True
+
+    def _func_sample() -> _Sample:
+        return _Sample()
+
+    def _func_none() -> None:
+        pass
+
+    ret_none = decorator_converter(_func_none)()
+    assert ret_none is None
+
+    ret_sample = decorator_converter(_func_sample)()
+    assert isinstance(ret_sample, _Sample)
+    assert ret_sample.method()
+
+
+def testconvert_positional_args_invalid_previous_positional_arg_names() -> None:
+    def _test_converter(previous_positional_arg_names: List[str], raise_error: bool) -> None:
+        decorator_converter = convert_positional_args(
+            previous_positional_arg_names=previous_positional_arg_names
+        )
+        assert callable(decorator_converter)
+
+        if raise_error:
+            with pytest.raises(AssertionError) as record:
+                decorator_converter(_sample_func)
+            assert str(record.value) == (
+                f"{set(previous_positional_arg_names)} is not a subset of"
+                f" {set(['a', 'b', 'c'])}"
+            )
```

Comment:
Nice checking the content of the message. I'm wondering however if the order (the printed string) can be guaranteed. 🤔

Reply:
Thank you for pointing out. Yes, the order of items in the printed string can vary and that should make this test flaky. I could not notice since this test did not fail a couple of times by luck.
Fixed in 3dd7917.

Reply:
Thanks!